packages/kokkos/core/src/impl/Kokkos_Core.cpp

```diff
@@ -93,21 +93,80 @@ setenv("MEMKIND_HBW_NODES", "1", 0);
   const int use_numa = args.num_numa;
 #endif
 #if defined( KOKKOS_ENABLE_CUDA ) || defined( KOKKOS_ENABLE_ROCM )
+  const std::string ctest_device_type = args.ctest_device_type;
   int use_gpu = args.device_id;
   const int ndevices = args.ndevices;
   const int skip_device = args.skip_device;
   // if the exact device is not set, but ndevices was given, assign round-robin using on-node MPI rank
-  if (use_gpu < 0 && ndevices >= 0) {
+  if (use_gpu < 0) {
     auto local_rank_str = std::getenv("OMPI_COMM_WORLD_LOCAL_RANK"); //OpenMPI
     if (!local_rank_str) local_rank_str = std::getenv("MV2_COMM_WORLD_LOCAL_RANK"); //MVAPICH2
     if (!local_rank_str) local_rank_str = std::getenv("SLURM_LOCALID"); //SLURM
-    if (local_rank_str) {
+
+    auto ctest_process_count_str = std::getenv("CTEST_PROCESS_COUNT"); //CTest
```

Comment:
Best practice would be to name the return type as `const char*`.  This would prevent code from attempting to change the string, which would [invoke undefined behavior](https://en.cppreference.com/w/cpp/utility/program/getenv).